src/sentry/grouping/result.py

```diff
@@ -0,0 +1,121 @@
+from dataclasses import dataclass
+from typing import TYPE_CHECKING, Any, Dict, Optional, Sequence, Union
+
+from sentry.utils.safe import get_path, safe_execute, set_path
+
+# TODO(3.8): This is a hack so we can get TypedDicts before 3.8
+if TYPE_CHECKING:
+    from mypy_extensions import TypedDict
+else:
+
+    def TypedDict(*args, **kwargs):
+        pass
+
+
+EventData = Dict[str, Any]
+EventMetadata = Dict[str, Any]
+
+
+TreeLabelPart = TypedDict(
+    "TreeLabelPart",
+    {
+        "function": str,
+        "package": str,
+        "is_sentinel": bool,
+        "is_prefix": bool,
+        "datapath": Sequence[Union[str, int]],
+    },
+)
+
+StrippedTreeLabelPart = TypedDict(
+    "StrippedTreeLabelPart",
+    {
+        "function": str,
+        "package": str,
+        "is_sentinel": bool,
+        "is_prefix": bool,
+    },
+)
+
+
+TreeLabel = Sequence[TreeLabelPart]
+StrippedTreeLabel = Sequence[StrippedTreeLabelPart]
+
+
+def _strip_tree_label(tree_label: TreeLabel) -> StrippedTreeLabel:
+    rv = []
+    for part in tree_label:
+        stripped_part: StrippedTreeLabelPart = dict(part)  # type: ignore
+        # TODO(markus): Remove more stuff here if we never use it in group
+        # title
+        stripped_part.pop("datapath", None)  # type: ignore
+        rv.append(stripped_part)
+
+    return rv
+
+
+def _write_tree_labels(tree_labels: Sequence[TreeLabel], event_data: EventData) -> None:
+    event_data["hierarchical_tree_labels"] = event_labels = []
+
+    for level, tree_label in enumerate(tree_labels):
+        event_labels.append(_strip_tree_label(tree_label))
+
+        for part in tree_label:
+            datapath = part["datapath"]
+            frame = get_path(event_data, *datapath)
+            if not frame:
+                raise ValueError("datapath not found in event")
+
+            if part.get("is_sentinel"):
+                set_path(frame, "data", "is_sentinel", value=True)
+
+            if part.get("is_prefix"):
+                set_path(frame, "data", "is_prefix", value=True)
+
+            prev_level = get_path(frame, "data", "min_grouping_level")
+            if not isinstance(prev_level, int) or level < prev_level:
```

Comment:
What happens here if after some change in grouping rules, an event is regrouped and the frame is now level >= 3 instead of level >= 2? Or is `min_grouping_level` cleared somewhere before `_write_tree_labels`?

Reply:
In other words, could a `min_grouping_level` persisted in a previous grouping run prevent a necessary update here?

Reply:
if you're changing grouping rules, old events are not rewritten or regrouped in any way

if you're changing the level, the `min_grouping_level` doesn't have to change

reprocessing allows you to reprocess events, but that works from a backup of the un-grouped event as it comes from relay.